knowledge-base/grid-inline-edit-form.md

```diff
@@ -0,0 +1,227 @@
+
+1. Define a [`DetailTemplate`]({%slug components/grid/features/hierarchy%}).
+1. Hide the hierarchy expand column with CSS.
+1. Use the [`SetStateAsync()` method]({%slug grid-state%}#setstateasync-examples) to enter and exit edit mode programmatically.
+
+>tip The Telerik UI for Blazor Form works with a cloned instance of the edited or added item to support cancellation. If you cancel the addition or update of an item, you need to delete it from the Grid data collection.   
+
+````CSHTML
+@using System.ComponentModel.DataAnnotations
+
+<TelerikGrid @ref="@GridRef"
+             Data="@GridData"
+             Class="no-expand-column">
+    <GridToolBarTemplate>
+        <GridCommandButton OnClick="@OnGridAddButtonClick" Icon="@SvgIcon.Plus">Add</GridCommandButton>
+    </GridToolBarTemplate>
+    <GridColumns>
+        <GridColumn Field="@nameof(SampleModel.Id)" />
+        <GridColumn Field="@nameof(SampleModel.Name)" />
+        <GridColumn Field="@nameof(SampleModel.Price)" DisplayFormat="{0:c2}" />
+        <GridColumn Field="@nameof(SampleModel.Quantity)" />
+        <GridColumn Field="@nameof(SampleModel.StartDate)" DisplayFormat="{0:d}" />
+        <GridCommandColumn>
+            @{
+                var item = (SampleModel)context;
+            }
+            <GridCommandButton OnClick="@( () => OnGridEditButtonClick(item) )" Icon="@SvgIcon.Pencil">Edit</GridCommandButton>
+        </GridCommandColumn>
+    </GridColumns>
+    <DetailTemplate>
+        <TelerikForm Model="@GridItemInEditMode"
+                     OnValidSubmit="@OnGridFormValidSubmit"
+                     Columns="2"
+                     ColumnSpacing="2em">
+            <FormValidation>
+                <DataAnnotationsValidator />
+            </FormValidation>
+            <FormButtons>
+                <TelerikButton ThemeColor="@ThemeConstants.Button.ThemeColor.Primary"
+                               Icon="@SvgIcon.Save">
+                    Save
+                </TelerikButton>
+                <TelerikButton ButtonType="@ButtonType.Button"
+                               OnClick="@OnGridFormCancel"
+                               Icon="@SvgIcon.Cancel">
+                    Cancel
+                </TelerikButton>
+            </FormButtons>
+        </TelerikForm>
+    </DetailTemplate>
+</TelerikGrid>
+
+<style>
+    .no-expand-column .k-hierarchy-col {
+        width: 0;
+    }
+
+    .no-expand-column .k-hierarchy-cell + .k-table-td,
+    .no-expand-column .k-hierarchy-cell + .k-table-th {
+        border-left-width: 0;
+    }
+
+    .no-expand-column .k-hierarchy-cell > * {
+        display: none;
+    }
+
+    .no-expand-column .k-detail-row:hover {
+        background-color: transparent;
+    }
+</style>
+
+@code {
+#nullable enable
+
+    private List<SampleModel> GridData { get; set; } = new();
+
+    private TelerikGrid<SampleModel> GridRef { get; set; } = null!;
```

Comment:
I used to code like this, but it's not the best way.

```suggestion
    private TelerikGrid<SampleModel>? GridRef { get; set; }
```

... and adjust the rest of the code. You will have to wrap `await GridRef.SetStateAsync(gridState);` in an `if` or use .